USERMANUAL.md

```diff
@@ -0,0 +1,599 @@
+
+```yaml
+rules:
+  GAP:
+    rule_type: CUSTOM_SQL_STATEMENT
+    custom_sql_arguments:
+      max_gap
+    params:
+      custom_sql_statement: |-
+        select $column, lag($column) over (order by $column ASC) as prev, $column - prev as gap 
+        from data
+        where gap <= max_gap
+        
+rule_bindings:
+  MAX_GAP:
+    entity_id: TEST_DATA
+    column_id: TIMESTAMP
+    row_filter_id: NONE
+    rule_ids:
+      - GAP:
+          max_gap: 3600
+```
+
+### Referential Integrity
+
+#### Reference against fixed set
+
+The fixed set of allowed values is hardcoded into the rule.
+
+```yaml
+rules:
+  ACCEPTED_CURRENCY:
+    rule_type: CUSTOM_SQL_EXPR
+    params:
+      custom_sql_expr: |-
+        $column in (‘USD’, ‘EUR’,  ‘CNY’)
+```
+
+#### Reference against subquery
+
+The subquery is coded into the `custom_sql_expr` field. This example assumes a reference table with a column with allowed values, but the query could be any that returns a list of values.
+
+```yaml
+rules:
+  REF_SUBQUERY:
+    rule_type: CUSTOM_SQL_EXPR
+    params:
+      custom_sql_expr: |-
+        $column in (select allowed_value from `<dataset_id>.<ref_table_id>`)
+```
+
+[Complete Example](https://github.com/GoogleCloudPlatform/cloud-data-quality/blob/docs_common_rules/docs/examples/reference_subquery.yaml)
+
+#### Foreign Key Constraint
+
+Check that all values in the referenced column exist in a column in another table (which can be interpreted as the primary key).
+
+This is a special case of the previous example.
+
+```yaml
+rules:
+  FOREIGN_KEY_VALID:
+    rule_type: CUSTOM_SQL_EXPR
+    params:
+      custom_sql_expr: |-
+        $column in (select distinct unique_key from `<dataset_id>.<pk_table_id>`)
+```
+
+### Row-by-row Dataset Comparison
+
+#### Equality of two columsn
+
+In  this test we check that two columns have an equal set of values. In this example we do not consider the ordering, meaning that if one column has `[a, a, b, b, c]` and another has `[a, b, c, a, b]`, then we will consider them equal.
+
+```yaml
+rules:
+  EQUAL_COLUMNS:
+    rule_type: CUSTOM_SQL_STATEMENT
+     custom_sql_arguments:
+ref_table_id
+ref_column_id
+    params:
+      custom_sql_statement: |-
+      with t1 as (
+        select $column as col, count(*) as t1_n 
+        from data
+        group by $column),
+     t2 as (
+        select $ref_column_id as col, count(*) as t2_n 
+        from $ref_table_id
+        group by $ref_column_id
+    )
+   select t1.*, t2.*
+   from t1 
+   outer join t2  using(col)
+   having not t1_n = t2_n 
+
+rule_bindings:
+  EQUAL_COLUMNS:
+    entity_id: TEST_DATA
+    column_id: TX_ID
+    row_filter_id: NONE
+    rule_ids:
+      - EQUAL_COLUMNS:
+          ref_table_id: <reference_table_id>
+          ref_column_id: <reference_column_id>
+```
+
+#### Equality of two columns (special case: columns do not contain repeated values)
```

Comment:
just a thought - given we've got so many examples of custom rules already, would it be a quick win to include them into the standard library?

Reply:
Consider for the backlog